# Exercicis per practicar amb regressions.
# Ejercicios por practicar con regresiones.
# Exercises to practice with regressions.

- Exercici 1

Crea almenys dos models de regressió diferents per intentar predir el millor possible el preu de les vivendes (MEDV) de l'arxiu adjunt.

- Ejercicio 1

Crea al menos dos modelos de regresión distintos para intentar predecir lo mejor posible el precio de las viviendas (MEDV) del archivo adjunto.

- Exercise 1

Create at least two different regression models to try to best predict the house price (MEDV) in the attached file.

In [1]:
from google.colab import drive
drive.mount('/content/gdrive')

Mounted at /content/gdrive


In [2]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import scipy

1. Title: Boston Housing Data

2. Sources:
   (a) Origin:  This dataset was taken from the StatLib library which is maintained at Carnegie Mellon University.
   (b) Creator:  Harrison, D. and Rubinfeld, D.L. 'Hedonic prices and the demand for clean air', J. Environ. Economics & Management, vol.5, 81-102, 1978.
   (c) Date: July 7, 1993

3. Past Usage:
   -   Used in Belsley, Kuh & Welsch, 'Regression diagnostics ...', Wiley, 1980.   N.B. Various transformations are used in the table on pages 244-261.
  -  Quinlan,R. (1993). Combining Instance-Based and Model-Based Learning. In Proceedings on the Tenth International Conference of Machine Learning, 236-243, University of Massachusetts, Amherst. Morgan Kaufmann.

4. Relevant Information:
Concerns housing values in suburbs of Boston.

5. Number of Instances: 506

6. Number of Attributes:
13 continuous attributes (including "class" attribute "MEDV"), 1 binary-valued attribute.

7. Attribute Information:

In [3]:
df = pd.read_csv('/content/gdrive/MyDrive/DS/S6.Algoritmes d’aprenentatge supervisat Regressió/housing data.csv', sep=',',header=None)

In [ ]:
df.head(3)

0     1     2   3      4      5     6       7   8      9     10  \
0  0.00632  18.0  2.31   0  0.538  6.575  65.2  4.0900   1  296.0  15.3   
1  0.02731   0.0  7.07   0  0.469  6.421  78.9  4.9671   2  242.0  17.8   
2  0.02729   0.0  7.07   0  0.469  7.185  61.1  4.9671   2  242.0  17.8   

       11    12    13  
0  396.90  4.98  24.0  
1  396.90  9.14  21.6  
2  392.83  4.03  34.7

In [9]:
cols = ['CRIM', 'ZN', 'INDUS', 'CHAS', 'NOX', 'RM', 'AGE', 'DIS', 'RAD', 'TAX', 'PTRATIO', 'B', 'LSTAT', 'MEDV']
df.columns = cols
df.head(3)

CRIM    ZN  INDUS  CHAS    NOX     RM   AGE     DIS  RAD    TAX  \
0  0.00632  18.0   2.31     0  0.538  6.575  65.2  4.0900    1  296.0   
1  0.02731   0.0   7.07     0  0.469  6.421  78.9  4.9671    2  242.0   
2  0.02729   0.0   7.07     0  0.469  7.185  61.1  4.9671    2  242.0   

   PTRATIO       B  LSTAT  MEDV  
0     15.3  396.90   4.98  24.0  
1     17.8  396.90   9.14  21.6  
2     17.8  392.83   4.03  34.7

In [5]:
import sklearn
from sklearn.model_selection import train_test_split

In [10]:
X = df.drop('MEDV', axis=1)
y = df['MEDV']

In [7]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [11]:
from sklearn.linear_model import LinearRegression

In [19]:
# Crear el modelo de regresión lineal
model = LinearRegression()

# Entrenar el modelo con el conjunto de entrenamiento
model.fit(X_train, y_train)

# Realizar predicciones en el conjunto de prueba
y_pred_linear = model.predict(X_test)

In [14]:
from sklearn.ensemble import RandomForestRegressor

In [15]:
# Crear y entrenar un modelo de regresión RandomForest
model_rf = RandomForestRegressor(random_state=42)
model_rf.fit(X_train, y_train)

# Realizar predicciones con el modelo RandomForest
y_pred_rf = model_rf.predict(X_test)


In [22]:
from sklearn.linear_model import Ridge

In [23]:
# Crear el modelo de regresión Ridge
ridge_model = Ridge(alpha=1.0)

# Entrenar el modelo con los datos de entrenamiento
ridge_model.fit(X_train, y_train)

# Realizar predicciones en los datos de prueba
y_pred_ridge = ridge_model.predict(X_test)


- Exercici 2

Compara’ls en base al MSE i al R2.

- Ejercicio 2

Compárales en base al MSE y al R2.

- Exercise 2

Compare them based on MSE and R2.


In [17]:
from sklearn.metrics import r2_score
from sklearn.metrics import mean_squared_error

In [20]:
# Calcular el MSE y R2 del modelo de regresión lineal
mse_linear = mean_squared_error(y_test, y_pred_linear)
r2_linear = r2_score(y_test, y_pred_linear)
print("MSE (Linear Regression):", mse_linear)
print("R2 (Linear Regression):", r2_linear)


MSE (Linear Regression): 24.291119474973478
R2 (Linear Regression): 0.6687594935356326


Un valor de R2 proper a 1 indica que el model és capaç d'explicar una gran proporció de la variabilitat en les dades de destí. En aquest cas, el valor de 0.67 suggereix que el model de regressió lineal explica aproximadament el 67% de la variabilitat en els preus d'habitatges (MEDV). Tanmateix, el MSE de 24.29 indica que encara hi ha un error significatiu en les prediccions del model.

Un valor de R2 cercano a 1 indica que el modelo es capaz de explicar una gran proporción de la variabilidad en los datos de destino. En este caso, el valor de 0.67 sugiere que el modelo de regresión lineal explica aproximadamente el 67% de la variabilidad en los precios de viviendas (MEDV). Sin embargo, el MSE de 24.29 indica que todavía hay un error significativo en las predicciones del modelo.

A value of R2 close to 1 indicates that the model is capable of explaining a large proportion of the variability in the target data. In this case, the value of 0.67 suggests that the linear regression model explains approximately 67% of the variability in housing prices (MEDV). However, the MSE of 24.29 indicates that there is still a significant error in the model's predictions.

In [21]:
# Calcular el MSE y R2 del modelo RandomForest
mse_rf = mean_squared_error(y_test, y_pred_rf)
r2_rf = r2_score(y_test, y_pred_rf)
print("MSE (RandomForest Regression):", mse_rf)
print("R2 (RandomForest Regression):", r2_rf)

MSE (RandomForest Regression): 7.901513892156864
R2 (RandomForest Regression): 0.8922527442109116


En aquest cas, el valor de 0.89 suggereix que el model de regressió de Random Forest explica aproximadament el 89% de la variabilitat en els preus d'habitatges (MEDV). A més, el MSE de 7.90 és considerablement més baix que el del model de regressió lineal, la qual cosa indica que el model de Random Forest té un millor ajust a les dades i fa prediccions més precises en comparació amb el model de regressió lineal.

En este caso, el valor de 0.89 sugiere que el modelo de regresión de Random Forest explica aproximadamente el 89% de la variabilidad en los precios de viviendas (MEDV). Además, el MSE de 7.90 es considerablemente más bajo que el del modelo de regresión lineal, lo que indica que el modelo de Random Forest tiene un mejor ajuste a los datos y realiza predicciones más precisas en comparación con el modelo de regresión lineal.

In this case, the value of 0.89 suggests that the Random Forest regression model explains approximately 89% of the variability in housing prices (MEDV). Furthermore, the MSE of 7.90 is considerably lower than that of the linear regression model, indicating that the Random Forest model has a better fit to the data and makes more accurate predictions compared to the linear regression model.

In [24]:
# Calcular el MSE y el R2
mse_ridge = mean_squared_error(y_test, y_pred_ridge)
r2_ridge = r2_score(y_test, y_pred_ridge)

print("MSE (Ridge Regression):", mse_ridge)
print("R2 (Ridge Regression):", r2_ridge)

MSE (Ridge Regression): 24.477191227708644
R2 (Ridge Regression): 0.6662221670168523


El model de regressió Ridge també té un R2 similar al model de regressió lineal, indicant que explica al voltant del 66,6% de la variabilitat en els preus de les vivendes. No obstant això, l'MSE és similar al del model de regressió lineal, la qual cosa significa que les prediccions del model de regressió Ridge també tenen un error significatiu en comparació amb els valors reals.

El modelo de regresión Ridge también tiene un R2 similar al modelo de regresión lineal, indicando que explica alrededor del 66.6% de la variabilidad en los precios de viviendas. Sin embargo, el MSE es similar al del modelo de regresión lineal, lo que significa que las predicciones del modelo de regresión Ridge también tienen un error significativo en comparación con los valores reales.

The Ridge regression model also has a similar R2 to the linear regression model, indicating that it explains around 66.6% of the variability in housing prices. However, the MSE is similar to that of the linear regression model, which means that the predictions of the Ridge regression model also have a significant error compared to the actual values.

- Exercici 3

Entrena’ls utilitzant els diferents paràmetres que admeten per intentar millorar-ne la predicció.

- Ejercicio 3

Entrenalos utilizando los diferentes parámetros que admiten para intentar mejorar su predicción.

- Exercise 3

Train them using the different parameters they support to try to improve their prediction.


EX1

- Exercici 4

Compara el seu rendiment emprant l’aproximació traint/test o emprant totes les dades (validació interna).

- Ejercicio 4

Compara su rendimiento utilizando la aproximación traicionando/test o utilizando todos los datos (validación interna).

- Exercise 4

Compare your performance using the traint/test approach or using all data (internal validation).



In [43]:
from sklearn.model_selection import cross_val_predict
from sklearn.model_selection import cross_val_score

# Crear modelos con diferentes parámetros
lr_model = LinearRegression()
ridge_model = Ridge(alpha=0.1)
rf_model = RandomForestRegressor(n_estimators=100, max_depth=10, random_state=42)

# Realizar validación cruzada y obtener predicciones
y_pred_lr = cross_val_predict(lr_model, X_train, y_train, cv=5)
y_pred_ridge = cross_val_predict(ridge_model, X_train, y_train, cv=5)
y_pred_rf = cross_val_predict(rf_model, X_train, y_train, cv=5)

# Evaluar los modelos con validación cruzada
mse_lr_scores = cross_val_score(lr_model, X_train, y_train, cv=5, scoring='neg_mean_squared_error')
r2_lr_scores = cross_val_score(lr_model, X_train, y_train, cv=5, scoring='r2')
print(f"MSE (Linear Regression): {abs(mse_lr_scores.mean())}")
print(f"R2 (Linear Regression): {r2_lr_scores.mean()}")

mse_ridge_scores = cross_val_score(ridge_model, X_train, y_train, cv=5, scoring='neg_mean_squared_error')
r2_ridge_scores = cross_val_score(ridge_model, X_train, y_train, cv=5, scoring='r2')
print(f"MSE (Ridge Regression): {abs(mse_ridge_scores.mean())}")
print(f"R2 (Ridge Regression): {r2_ridge_scores.mean()}")

mse_rf_scores = cross_val_score(rf_model, X_train, y_train, cv=5, scoring='neg_mean_squared_error')
r2_rf_scores = cross_val_score(rf_model, X_train, y_train, cv=5, scoring='r2')
print(f"MSE (Random Forest Regression): {abs(mse_rf_scores.mean())}")
print(f"R2 (Random Forest Regression): {r2_rf_scores.mean()}")


MSE (Linear Regression): 23.654418270304568
R2 (Linear Regression): 0.7244357342240269
MSE (Ridge Regression): 23.677776898980472
R2 (Ridge Regression): 0.7241044173937535
MSE (Random Forest Regression): 14.940847065407947
R2 (Random Forest Regression): 0.8251901857728022


En aquesta comparació, és evident que el model de Regressió Random Forest té el millor rendiment en termes tant de l'error mitjà quadràtic (MSE) com dels valors del coeficient de determinació (R2). Els valors més baixos de MSE i els valors més alts de R2 pel model Random Forest suggereixen que ofereix una millor adaptació als dades i produeix prediccions més precises en comparació amb els models de Regressió Lineal i Regressió Ridge.

En esta comparación, es evidente que el modelo de Regresión Random Forest tiene el mejor rendimiento en términos tanto del error cuadrático medio (MSE) como de los valores del coeficiente de determinación (R2). Los valores más bajos de MSE y los valores más altos de R2 para el modelo Random Forest sugieren que proporciona un mejor ajuste a los datos y produce predicciones más precisas en comparación con los modelos de Regresión Lineal y Regresión Ridge.

In this comparison, it's evident that the Random Forest Regression model is performing the best in terms of both MSE (Mean Squared Error) and R2 (Coefficient of Determination) scores. The lower MSE and higher R2 values for the Random Forest model suggest that it provides a better fit to the data and produces more accurate predictions compared to the Linear Regression and Ridge Regression models.

- Exercici 5

No facis servir la variable del nombre d'habitacions (RM) a l’hora de fer prediccions.

- Ejercicio 5

No utilices la variable del número de habitaciones (RM) a la hora de realizar predicciones.

- Exercise 5

Do not use the number of rooms (RM) variable when making predictions.


In [46]:
# Excluir la variable 'RM'
X = df.drop(['MEDV', 'RM'], axis=1)
y = df['MEDV']

# Dividir los datos en conjuntos de entrenamiento y prueba
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Crear y entrenar el modelo de Regresión Lineal
modelo_lr = LinearRegression()
modelo_lr.fit(X_train, y_train)

# Crear y entrenar el modelo de Regresión Ridge
modelo_ridge = Ridge()
modelo_ridge.fit(X_train, y_train)

# Crear y entrenar el modelo de Regresión de Bosque Aleatorio
modelo_rf = RandomForestRegressor()
modelo_rf.fit(X_train, y_train)

# Realizar predicciones
predicciones_lr = modelo_lr.predict(X_test)
predicciones_ridge = modelo_ridge.predict(X_test)
predicciones_rf = modelo_rf.predict(X_test)

# Calcular los puntajes de MSE y R2
mse_lr = mean_squared_error(y_test, predicciones_lr)
r2_lr = r2_score(y_test, predicciones_lr)

mse_ridge = mean_squared_error(y_test, predicciones_ridge)
r2_ridge = r2_score(y_test, predicciones_ridge)

mse_rf = mean_squared_error(y_test, predicciones_rf)
r2_rf = r2_score(y_test, predicciones_rf)

# Imprimir los resultados
print(f"MSE (Regresión Lineal): {mse_lr}")
print(f"R2 (Regresión Lineal): {r2_lr}")

print(f"MSE (Regresión Ridge): {mse_ridge}")
print(f"R2 (Regresión Ridge): {r2_ridge}")

print(f"MSE (Regresión de Random Forest): {mse_rf}")
print(f"R2 (Regresión de Random Forest): {r2_rf}")

MSE (Regresión Lineal): 22.133246888189877
R2 (Regresión Lineal): 0.6981848483146986
MSE (Regresión Ridge): 22.502067392105385
R2 (Regresión Ridge): 0.6931555086567514
MSE (Regresión de Random Forest): 12.85440306862745
R2 (Regresión de Random Forest): 0.8247137606343717


En aquesta comparació, és evident que el model de Regressió de Random Forest té el millor rendiment en termes de puntuacions de MSE (Error Mitjà Quadrat) i R2 (Coeficient de Determinació). Els valors més baixos de MSE i els valors més alts de R2 pel model de Bosc Aleatori suggereixen que proporciona un ajust millor als dades i genera prediccions més precises en comparació amb els models de Regressió Lineal i Regressió Ridge.

En esta comparación, es evidente que el modelo de Regresión de Random Forest tiene el mejor rendimiento en términos de puntajes de MSE (Error Cuadrático Medio) y R2 (Coeficiente de Determinación). Los valores más bajos de MSE y los valores más altos de R2 para el modelo de Bosque Aleatorio sugieren que proporciona un ajuste mejor a los datos y genera predicciones más precisas en comparación con los modelos de Regresión Lineal y Regresión Ridge.

In this comparison, it's evident that the Random Forest Regression model performs the best in terms of both MSE (Mean Squared Error) and R2 (Coefficient of Determination) scores. The lower MSE and higher R2 values for the Random Forest model suggest that it provides a better fit to the data and produces more accurate predictions compared to the Linear Regression and Ridge Regression models.